File: src/policy.ts
export interface Band { no: number; yes: number }
export function probability(p: number): number {
  if (!Number.isFinite(p) || p < 0 || p > 1) throw new RangeError("probability must be in [0,1]");
  return p;
}
export function decide(p: number, band: Band): "no" | "uncertain" | "yes" {
  if (!Number.isFinite(band.no) || !Number.isFinite(band.yes) || band.no < 0 || band.no >= band.yes || band.yes > 1) throw new RangeError("invalid band");
  return probability(p) < band.no ? "no" : p > band.yes ? "yes" : "uncertain";
}
export class Hysteresis<T> {
  private candidate: T | undefined;
  private count = 0;
  private current: T | undefined;
  constructor(private readonly ticks: number) {
    if (!Number.isInteger(ticks) || ticks < 1) throw new RangeError("ticks must be a positive integer");
  }
  step(value: T): T | undefined {
    if (Object.is(value, this.candidate)) this.count++;
    else { this.candidate = value; this.count = 1; }
    if (this.count >= this.ticks) this.current = value;
    return this.current;
  }
  reset(): void { this.candidate = undefined; this.current = undefined; this.count = 0; }
}
export class Refractory {
  private last = -Infinity;
  constructor(private readonly windowMs: number) {
    if (!Number.isFinite(windowMs) || windowMs < 0) throw new RangeError("windowMs must be non-negative");
  }
  fire(nowMs: number): boolean {
    if (!Number.isFinite(nowMs) || nowMs < this.last) return false;
    if (nowMs - this.last < this.windowMs) return false;
    this.last = nowMs;
    return true;
  }
}
export function composite(values: Record<string, number>, weights: Record<string, number>): number {
  const keys = Object.keys(weights);
  if (!keys.length) throw new RangeError("weights required");
  const total = keys.reduce((sum, key) => {
    const weight = weights[key];
    if (weight === undefined || !Number.isFinite(weight) || weight < 0) throw new RangeError("invalid weight");
    if (values[key] === undefined) throw new TypeError(`missing value: ${key}`);
    return sum + weight;
  }, 0);
  if (total <= 0) throw new RangeError("positive total weight required");
  return Math.max(0, Math.min(1, keys.reduce((sum, key) => sum + probability(values[key]!) * weights[key]!, 0) / total));
}
export function gate<T>(choice: { choice: T; confidence: number }, minConfidence: number): T | null {
  probability(minConfidence);
  return probability(choice.confidence) >= minConfidence ? choice.choice : null;
}

File: python/src/reachy_jev/policy.py
"""Deterministic policy primitives; no model or robot calls."""

from __future__ import annotations

import math
from typing import TypeVar

T = TypeVar("T")


def probability(value: float) -> float:
    if (
        isinstance(value, bool)
        or not isinstance(value, (float, int))
        or not math.isfinite(value)
        or not 0 <= value <= 1
    ):
        raise ValueError("probability must be in [0,1]")
    return value


def decide(value: float, *, no: float, yes: float) -> str:
    if not 0 <= no < yes <= 1:
        raise ValueError("invalid band")
    probability(value)
    return "no" if value < no else "yes" if value > yes else "uncertain"


class Hysteresis:
    def __init__(self, ticks: int) -> None:
        if isinstance(ticks, bool) or not isinstance(ticks, int) or ticks < 1:
            raise ValueError("ticks must be a positive integer")
        self.ticks = ticks
        self.reset()

    def step(self, value: T) -> T | None:
        if value == self._candidate:
            self._count += 1
        else:
            self._candidate = value
            self._count = 1
        if self._count >= self.ticks:
            self._current = value
        return self._current

    def reset(self) -> None:
        self._candidate = None
        self._current = None
        self._count = 0


class Refractory:
    def __init__(self, window_ms: float) -> None:
        if (
            not isinstance(window_ms, (int, float))
            or isinstance(window_ms, bool)
            or not math.isfinite(window_ms)
            or window_ms < 0
        ):
            raise ValueError("window_ms must be non-negative")
        self.window_ms = window_ms
        self._last = -math.inf

    def fire(self, now_ms: float) -> bool:
        if (
            not isinstance(now_ms, (int, float))
            or isinstance(now_ms, bool)
            or not math.isfinite(now_ms)
            or now_ms < self._last
            or now_ms - self._last < self.window_ms
        ):
            return False
        self._last = now_ms
        return True


def composite(values: dict[str, float], weights: dict[str, float]) -> float:
    if not weights:
        raise ValueError("weights required")
    total = 0.0
    weighted = 0.0
    for key, weight in weights.items():
        if (
            not isinstance(weight, (int, float))
            or isinstance(weight, bool)
            or not math.isfinite(weight)
            or weight < 0
        ):
            raise ValueError("invalid weight")
        if key not in values:
            raise KeyError(f"missing value: {key}")
        total += weight
        weighted += probability(values[key]) * weight
    if total <= 0:
        raise ValueError("positive total weight required")
    return max(0.0, min(1.0, weighted / total))


def gate(choice: T, confidence: float, min_confidence: float) -> T | None:
    probability(min_confidence)
    return choice if probability(confidence) >= min_confidence else None
